asf_online_data_exploration/pipeline/data_collection/the_guardian.py

```diff
@@ -0,0 +1,151 @@
+"""
+Script to collect data from The Guardian Open Platform on a topic of choice using the content endpoint.
+
+A developer is allowed to make up to 500 calls per day and 1 call per second.
+"""
+
+import requests
+import json
+import time
+import random
+import json
+import os
+from datetime import datetime
+from asf_online_data_exploration.config.data_collection_parameters import (
+    heating_technologies_ruleset_guardian,
+    query_parameters_guardian,
+    dates_for_guardian_collection,
+)
+from asf_online_data_exploration.utils.data_collection_utils import dictionary_to_s3
+
+ENDPOINT_URL = "https://content.guardianapis.com/search?"
+S3_BUCKET = "asf-online-data-exploration"
+DATA_COLLECTION_FOLDER = "inputs/data_collection/the_guardian/"
+
+
+def define_endpoint_url(api_key, query_params):
+    """
+    Defines endpoint url starting from the base content endpoint URL and
+    using the api_key and query parameters.
+    Args:
+        api_key: API key credentials
+        query_params: query parameters
+    Returns:
+        Endpoint URL ready for API call
+    """
+    url = f"{ENDPOINT_URL}&api-key={api_key}"
+    for key in query_params.keys():
+        url = url + f"&{key}={query_params[key]}"
+    return url
+
+
+def connect_to_endpoint(endpoint_url: str) -> json:
+    """
+    Connects to the endpoint and requests data.
+    Returns a json with The Guardian data if a 200 status code is yielded.
+    Programme stops if there is a problem with the request and sleeps
+    if there is a temporary problem accessing the endpoint.
+
+    Args:
+        endpoint_url: url to endpoint
+    """
+    response = requests.request("GET", url=endpoint_url)
+    response_status_code = response.status_code
+    if response_status_code != 200:
+        if response_status_code >= 400 and response_status_code < 500:
+            raise Exception(
+                "Cannot get data, the program will stop!\nHTTP {}: {}".format(
+                    response_status_code, response.text
+                )
+            )
+
+        sleep_seconds = random.randint(5, 60)
+        print(
+            "Cannot get data, your program will sleep for {} seconds...\nHTTP {}: {}".format(
+                sleep_seconds, response_status_code, response.text
+            )
+        )
+        time.sleep(sleep_seconds)
+        return connect_to_endpoint(endpoint_url)
+    # sleep 1 second not to surpass rate limit
+    time.sleep(1)
+    return response.json()["response"]
+
+
+def collect_and_process_guardian_data(
+    api_key, rules, query_params, s3_bucket: str, s3_folder: str
```

Comment:
Just a thought: it might be nice to have `s3_bucket` as an optional parameter - and if it's not given then the output is saved locally to `s3_folder` (maybe rename to "folder_name" or something)? Just thought this because I wanted to download some data but then realised it might mess up your s3 bucket if I did!

Reply:
Great suggestion, thanks @lizgzil ! :)
I added that option to both data collection scripts, and also added new tests to`testing_the_guardian.py` and `testing_recent_search_twitter.py`. Hopefully the way I implemented it makes sense to you. So now if you re-run:

- `pytest asf_online_data_exploration/pipeline/data_collection/tests/testing_the_guardian.py` you should have data in your local inputs folder: `/inputs/data_collection/the_guardian/test/`
- `pytest asf_online_data_exploration/pipeline/data_collection/tests/testing_recent_search_twitter.py` you should have data in your local inputs folder: `/inputs/data_collection/recent_search_twitter/test/` (should only work until the 13th of Feb, so no need to worry about this)

Could you please check if the tests still pass and that you indeed end up with data in your local inputs folder?

Reply:
great! yep both tests pass 👍

Reply:
thanks @lizgzil :D !